### DISCUSSION TITLE
kdevtmpfsi && permission denied to set role \"anon\"

### DISCUSSION THREAD

**Discussion by macro0000** [PostgREST/postgrest#2194] [Q&A] [answered]
permission denied to set role \"anon\"

This problem occurred after I killed kdevtmpfsi and cleared all contents under 'tmp'

**Comment by steve-chavez** [comment, marked as answer]
Are you sure these events are related?

**Reply by fusionbeam** [reply]
I have a similar issue.
Strangely enough in my case it only happens to about 25% of requests.
How can this be?

**Reply by steve-chavez** [reply]
@fusionbeam Hm, what's your issue? The `permission denied to set role "anon"` error happening for those requests?

**Reply by fusionbeam** [reply]
@steve-chavez 
I have run into a nasty issue after a postgres upgrade.
About 25% of requests fail with error: permission denied to set role "anon" (or some other specific user role).
using Postgrest 9, PostreSQL 13.6.
Any wild guess on why would some requests work and others not? Same schema, same user, same authorization header, even the same request if repeated works 2 out of 3.

**Reply by steve-chavez** [reply]
> About 25% of requests fail with error: permission denied to set role "anon" (or some other specific user role).

Hm, I think that could only happen if the `authenticator` is missing the privilege to switch to anon, so missing a:

```sql
grant anon to authenticator;
```

(As mentioned [here](https://postgrest.org/en/stable/auth.html#permissions))

> I have run into a nasty issue after a postgres upgrade.

Since roles are global, perhaps the upgrade did something to them(like their privileges got reset), hmm.. maybe try rerunning the GRANTs.

**Reply by fusionbeam** [reply]
@steve-chavez 
NOTICE:  role "authenticator" is already a member of role "anon"
which makes sense since 2 times out of 3 the same api request works.

Also tried running select set_config('role','anon',true) and work every time.

I have also dropped the db, restore it and I get the same behaviour.

Interestingly, I am running two schemas, each exposed by Postgrest daemonised by systemd.
Only one of the schema has this issue.

**Reply by wolfgangwalther** [reply]
> which makes sense since 2 times out of 3 the same api request works.

Are you using a loader balancer like pgbouncer or something?

**Reply by fusionbeam** [reply]
no load balancer, single postgrest server wrapped by nginx

**Reply by fusionbeam** [reply]
for what is worth, I've managed to find the problem.
In one of the RPC functions I was doing a SET ROLE/SET SESSION AUTHORIZATION and then RESET ROLE before returning.

I think RESET ROLE was causing the problems in conjunction with thread pooling.

**Reply by wolfgangwalther** [reply]
When you're messing with `SET SESSION AUTHORIZATION` you are doing something wrong. The [docs](https://www.postgresql.org/docs/current/sql-set-session-authorization.html):

> The session user identifier can be changed only if the initial session user (the authenticated user) had the superuser privilege. Otherwise, the command is accepted only if it specifies the authenticated user name.

That means you're running PostgREST with superuser privileges?

You should **not** do that.

As for `SET ROLE` you should always use that as `SET LOCAL ROLE`. This will set the role only for the current transaction - and you don't need to clean things up afterwards with `RESET ROLE`.

**Reply by macro0000** [reply]
`kdevtmpfsi ` maliciously changed my permission. I was eager to recover and chose to reset. Thank you. Your suggestion is very useful
